File: 207-course-schedule/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Queue<Integer> q= new LinkedList<>(); 
        int[] dis= new int[numCourses];
        ArrayList<ArrayList<Integer>> adj= new ArrayList<>(); 
        for(int i=0;i<numCourses;i++) 
        { 
            adj.add(new ArrayList<Integer>()); 
        } 
        for(int[] edg:prerequisites) { 
            int u=edg[0]; 
            int v=edg[1]; 
            adj.get(v).add(u); 
            dis[u]++;
        }

        for(int i=0;i<numCourses;i++)
        {
            if(dis[i]==0)
            {
                q.add(i);
            }
        }

        while(!q.isEmpty())
        {
            int val=q.poll();
            for(int edg:adj.get(val))
            {
                if(dis[edg]!=0)
                {
                    dis[edg]--;
                }
                if(dis[edg]==0)
                {
                    q.add(edg);
                }
            }
        }
        for(int i=0;i<dis.length;i++)
        {
            if(dis[i]!=0)
            {
                return false;
            }
        }
        return true;

    }
}

File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& arr) {
        // kahn's algo
        vector<vector<int>> adj(n);
        vector<int> inDegree(n,0);

        for(auto edg: arr){
            int u=edg[0];
            int v=edg[1];
            inDegree[v]++;
            adj[u].push_back(v);
        }

        queue<int> q;

        for(int i=0;i<n;i++){
            if(inDegree[i]==0){
                q.push(i);
            }
        }
        
        while(!q.empty()){
            int node=q.front();
            q.pop();

            for(auto nbr: adj[node]){
                inDegree[nbr]--;
                if(inDegree[nbr]==0){
                    q.push(nbr);
                }
            }
        }

        for(int i=0;i<n;i++){
            if(inDegree[i]!=0)
            return false;
        }
        return true;
    }
};